src/main/java/org/opensearch/security/resources/migrate/MigrateResourceSharingInfoApiAction.java

```diff
@@ -0,0 +1,334 @@
+
+        List<SourceDoc> results = new ArrayList<>();
+
+        // 1) configure a 1-minute scroll
+        Scroll scroll = new Scroll(TimeValue.timeValueMinutes(1L));
+        SearchRequest searchRequest = new SearchRequest(sourceIndex).scroll(scroll)
+            .source(
+                new SearchSourceBuilder().query(QueryBuilders.matchAllQuery()).size(1_000)        // batch size per scroll “page”
+            );
+
+        // 2) execute first search
+        SearchResponse searchResponse = client.search(searchRequest).actionGet();
+        String scrollId = searchResponse.getScrollId();
+
+        // 3) page through until no hits
+        while (true) {
+            SearchHit[] hits = searchResponse.getHits().getHits();
+            if (hits == null || hits.length == 0) {
+                break;
+            }
+            for (SearchHit hit : hits) {
+                JsonNode rec = Utils.toJsonNode(hit.getSourceAsString());
+                String id = hit.getId();
+                String username = rec.at(userNamePath.startsWith("/") ? userNamePath : ("/" + userNamePath)).asText(null);
+
+                // backend_roles as an actual array
+                JsonNode backendRolesNode = rec.at(backendRolesPath.startsWith("/") ? backendRolesPath : ("/" + backendRolesPath));
+                List<String> backendRoles = new ArrayList<>();
+                if (backendRolesNode.isArray()) {
+                    for (JsonNode br : backendRolesNode) {
+                        backendRoles.add(br.asText());
+                    }
+                }
+
+                results.add(new SourceDoc(id, username, backendRoles));
+            }
+            // 4) fetch next batch
+            SearchScrollRequest scrollRequest = new SearchScrollRequest(scrollId).scroll(scroll);
+            searchResponse = client.searchScroll(scrollRequest).actionGet();
+            scrollId = searchResponse.getScrollId();
+        }
+
+        // 5) clear the scroll context to free resources
+        ClearScrollRequest clear = new ClearScrollRequest();
+        clear.addScrollId(scrollId);
+        client.clearScroll(clear).actionGet();
+
+        return ValidationResult.success(Triple.of(sourceIndex, defaultAccessLevel, results));
+    }
+
+    /**
+     * This method:
+     *      1. Parses existing sharing info to a new ResourceSharing records
+     *      2. Indexes the new record into corresponding resource-sharing index
+     */
+    private ValidationResult<String> createNewSharingRecords(Triple<String, String, List<SourceDoc>> sourceInfo) throws IOException {
+        AtomicInteger migratedCount = new AtomicInteger();
+        AtomicInteger skippedNoUser = new AtomicInteger();
+        AtomicInteger failureCount = new AtomicInteger();
+
+        List<SourceDoc> docs = sourceInfo.getRight();
+        int total = docs.size();
+
+        CountDownLatch latch = new CountDownLatch(total);
```

Comment:
Why is `CountDownLatch` used in case?

Reply:
For tracking resource migration stats.

Reply:
But we are blocking on it in line 262, right? So, this is for synchronizing on completion?

I'd recommend to rename this to `pendingResources` or something similar in this case.